core/src/main/scala/chisel3/Aggregate.scala

```diff
@@ -633,13 +645,11 @@ abstract class Record(private[chisel3] implicit val compileOptions: CompileOptio
     case _ => false
   }
 
-  // NOTE: This sets up dependent references, it can be done before closing the Module
   private[chisel3] override def _onModuleClose: Unit = {
-    // Since Bundle names this via reflection, it is impossible for two elements to have the same
-    // identifier; however, Namespace sanitizes identifiers to make them legal for Firrtl/Verilog
-    // which can cause collisions
-    val _namespace = Namespace.empty
-    for ((name, elt) <- elements) { elt.setRef(this, _namespace.name(name, leadingDigitOk=true)) }
+    // This is usually done during binding, but these must still be set for unbound Records
```

Comment:
Is there an example of when this could be the case? If possible, it'd be nice for setElementRefs to only be called from one place?

Reply:
I would love to have it only call from one place; the reason I factored the code out into a function was because AFAICT that's not possible. The basic problem is that you have to do it at some point, and since users provide `def elements` in `Record`, there's not really a hook to do it. We could do it as part of `lazy val elements` in `Bundles`, but that doesn't solve the `Record` problem.

Reply:
Are we doing something with unbound Data (or maybe I'm missing something really basic here)?

Reply:
I was thinking about `val io` in compatibility modules, but there is `bindIoInPlace`, so maybe we don't need to bother setting the refs on module close. The fact that `io` in compatibility modules may not yet be bound *does* mean it's important that the naming not fail for aggregates without refs, they'll just get slightly worse prefixes for `io.subfield :=` connections in compatibility modules. We can still get good names if we do ref sending in `lazy val elements` inside `Bundle`, but then we need to make sure we still set the refs on binding for non-Bundle `Records`.

Reply:
I'd say this is answered pretty well in https://github.com/freechipsproject/chisel3/pull/1616#issuecomment-714011111. I've reverted the changes discussed in that comment and will leave those improvements for future work.